### PR TITLE
Closes #5992: Get rid of the cache tab

### LINKED ISSUE #5992: Get rid of the Cache Tab
**Is your feature request related to a problem? Please describe.**
This is a GitHub part of the enhancement that's described [here](https://www.notion.so/wpmedia/Get-rid-of-the-Cache-tab-e505d7b0ffc54cc8b0920cdfb4bdf853?pvs=4).

**Additional context**
Acceptanca Critera and details are present in the above document.

### PR DESCRIPTION
**Pull request by jeawhanlee**
## Description
This PR removes the cache tab as described in the [epic](https://www.notion.so/wpmedia/Get-rid-of-the-Cache-tab-e505d7b0ffc54cc8b0920cdfb4bdf853)

Fixes #5992 

## Type of change

- [x] Enhancement (non-breaking change which improves an existing functionality)

## Is the solution different from the one proposed during the grooming?
No

## How Has This Been Tested?
Automated & Manual Tests

# Checklist:

- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] My changes generate no new warnings
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] New and existing unit tests pass locally with my changes
- [ ] Any dependent changes have been merged and published in downstream modules

### CODE REVIEW

**Comment by piotrbak** [comment]
@jeawhanlee Where are we at with this PR?

**Comment by jeawhanlee** [comment]
> @jeawhanlee Where are we at with this PR?

@piotrbak Done, just awaiting assets from the Product Design team

File: inc/Engine/Admin/Settings/Page.php [hunk 1]
```diff
@@ -2251,4 +2214,52 @@ public function display_radio_options_sub_fields( $sub_fields ) {
...
+	 * @return void
+	 */
+	public function enable_separate_cache_files_mobile() : void {
+		if ( ! (bool) get_rocket_option( 'cache_mobile', 0 ) ) {
+			return;
+		}
+
+		$this->options->set( 'do_caching_mobile_files', 1 );
+		update_option( rocket_get_constant( 'WP_ROCKET_SLUG', 'wp_rocket_settings' ), $this->options->get_options() );
+	}
```
Comment (remyperona) [inline inc/Engine/Admin/Settings/Page.php]: Missing a check for the current version number vs the update version number.

We can also bail-out early if the option is already enabled

File: inc/classes/admin/class-options-data.php [hunk 2]
```diff
@@ -58,9 +58,39 @@ public function get( $key, $default = '' ) {
...
+
+				break;
+			case 'do_caching_mobile_files':
+				if ( isset( $this->options[ $key ] ) && rocket_get_config( 'do_caching_mobile_files' ) !== $this->options[ $key ] ) {
+					rocket_generate_config_file();
+				}
+
+				break;
+		}
+
```
Comment (remyperona) [inline inc/classes/admin/class-options-data.php]: I don't like this, it's introducing specific coupling to a class that is mainly generic. Why is this needed and what is the goal?

File: inc/functions/files.php [hunk 3]
```diff
@@ -52,9 +52,10 @@ function rocket_generate_advanced_cache_file( $advanced_cache = null ) {
  *
  * @since 2.0
  *
+ * @param bool $cache_mobile Mobile Cache config value.
  * @return array Names of all config files & The content that will be printed
  */
-function get_rocket_config_file() { // phpcs:ignore WordPress.NamingConventions.PrefixAllGlobals
+function get_rocket_config_file( $cache_mobile = true ) { // phpcs:ignore WordPress.NamingConventions.PrefixAllGlobals
```
Comment (remyperona) [inline inc/functions/files.php]: why is this needed?

**Review by remyperona — CHANGES_REQUESTED** [review]

File: inc/classes/admin/class-options-data.php (hunk 2 above)
Comment (jeawhanlee) [reply inc/classes/admin/class-options-data.php]: @Tabrisrp We are not regenerating the config file when the `pre_get_rocket_option` filter is used to disable mobile cache or separate cache for mobile, that was the reason for this. Do you see a better way to handle this?

File: inc/functions/files.php (hunk 3 above)
Comment (jeawhanlee) [reply inc/functions/files.php]: To disable the mobile cache options when the `pre_get_rocket_option` is fired

File: inc/Engine/Admin/Settings/Page.php [hunk 4]
```diff
@@ -2251,4 +2214,56 @@ public function display_radio_options_sub_fields( $sub_fields ) {
...
 		$this->render->render_fields( $sub_fields );
 	}
+
+	/**
+	 * Render mobile cache option.
+	 *
+	 * @return void
+	 */
+	public function display_mobile_cache_option() : void {
+		if ( (bool) get_rocket_option( 'cache_mobile', 0 ) ) {
```
Comment (remyperona) [inline inc/Engine/Admin/Settings/Page.php]:
```suggestion
		if ( (bool) $this->options->get( 'cache_mobile', 0 ) ) {
```

File: inc/Engine/Admin/Settings/Page.php [hunk 5]
```diff
@@ -2251,4 +2214,56 @@ public function display_radio_options_sub_fields( $sub_fields ) {
...
+		wp_send_json_success();
+	}
+
+	/**
+	 * Enable Separate cache files option on upgrade.
+	 *
+	 * @return void
+	 */
+	public function enable_separate_cache_files_mobile() : void {
+		if ( ! (bool) get_rocket_option( 'cache_mobile', 0 ) ) {
```
Comment (remyperona) [inline inc/Engine/Admin/Settings/Page.php]:
```suggestion
		if ( ! (bool) $this->options->( 'cache_mobile', 0 ) ) {
```

File: inc/Engine/Admin/Settings/Page.php [hunk 6]
```diff
@@ -2251,4 +2214,56 @@ public function display_radio_options_sub_fields( $sub_fields ) {
...
+			return;
+		}
+
+		if ( (bool) get_rocket_option( 'do_caching_mobile_files', 0 ) ) {
```
Comment (remyperona) [inline inc/Engine/Admin/Settings/Page.php]:
```suggestion
		if ( (bool) $this->options->get( 'do_caching_mobile_files', 0 ) ) {
```

File: inc/classes/admin/class-options-data.php (hunk 2 above)
Comment (remyperona) [reply inc/classes/admin/class-options-data.php]: It could be done in the callback hooked on this filter instead. Do we have use cases in the plugin?

**Review by remyperona — CHANGES_REQUESTED** [review]

File: inc/classes/admin/class-options-data.php (hunk 2 above)
Comment (jeawhanlee) [reply inc/classes/admin/class-options-data.php]: > It could be done in the callback hooked on this filter instead. Do we have use cases in the plugin?

@Tabrisrp I don't think we have use cases but since we are giving the users the freedom of disabling these options, there are chances, they won't be handling the regeneration of config file themselves except we document this during the use of these filters.

Comment (remyperona) [reply inc/classes/admin/class-options-data.php]: In my opinion this should be handled in the callback implementation with the documentation helping for this specific case. Usage of hooks is for developers/advanced users, so it makes the most sense to me.

**Review by CrochetFeve0251 — APPROVED** [review]

File: inc/Engine/Admin/Settings/Subscriber.php [hunk 7]
```diff
@@ -231,4 +233,37 @@ public function display_radio_options_sub_fields( $option_data ) {
...
+
+	/**
+	 * Enable Separate cache files for mobile devices on upgrade.
+	 *
+	 * @param string $new_version New plugin version.
+	 * @param string $old_version Previous plugin version.
+	 * @return void
+	 */
+	public function enable_separate_cache_files_mobile( $new_version, $old_version ) : void {
+		if ( version_compare( $old_version, '3.16', '>' ) ) {
```
Comment (remyperona) [inline inc/Engine/Admin/Settings/Subscriber.php:263]: This should be version 3.15 no?

Comment (jeawhanlee) [reply inc/Engine/Admin/Settings/Subscriber.php:263]: The milestone changed to 3.16

File: inc/Engine/Admin/Settings/Page.php [hunk 8]
```diff
@@ -1817,6 +1751,28 @@ private function addons_section() {
...
+					'type'              => 'one_click_addon',
+					'label'             => __( 'User Cache', 'rocket' ),
+					'logo'              => [
+						'url'    => WP_ROCKET_ASSETS_IMG_URL . 'icon-user-cache.svg',
+						'width'  => 152,
+						'height' => 135,
+					],
+					'title'             => __( 'If you need to create a dedicated set of cache files for each logged-in WordPress user, you must activate this add-on.', 'rocket' ),
+					// translators: %1$s = opening <a> tag, %2$s = closing </a> tag.
+					'description'       => sprintf( __( 'User cache is great when you have user-specific or restricted content on your website..<br>%1$sLearn more%2$s', 'rocket' ), '<a href="' . esc_url( $user_cache_beacon['url'] ) . '" data-beacon-article="' . esc_attr( $user_cache_beacon['id'] ) . '" target="_blank">', '</a>' ),
```
Comment (remyperona) [inline inc/Engine/Admin/Settings/Page.php]:
```suggestion
					'description'       => sprintf( __( 'User cache is great when you have user-specific or restricted content on your website.<br>%1$sLearn more%2$s', 'rocket' ), '<a href="' . esc_url( $user_cache_beacon['url'] ) . '" data-beacon-article="' . esc_attr( $user_cache_beacon['id'] ) . '" target="_blank">', '</a>' ),
```

File: inc/functions/files.php [hunk 9]
```diff
@@ -222,6 +222,19 @@ function rocket_generate_config_file() {
...
+ *
+ * @param string $config_name Name of a specific config/option.
+ * @return mixed
+ */
+function rocket_get_config( $config_name ) {
+	$container = apply_filters( 'rocket_container', null );
+	$config    = $container->get( 'config' );
+
+	return $config->get_config( $config_name );
+}
```
Comment (remyperona) [inline inc/functions/files.php]: I think this is not used anymore?

File: tests/Unit/inc/Engine/Admin/Settings/Page/enableMobileCache.php [hunk 10]
```diff
@@ -0,0 +1,88 @@
...
+
+use Mockery;
+use Brain\Monkey\Functions;
+use WP_Rocket\Engine\Admin\Database\Optimization;
+use WP_Rocket\Engine\Admin\Beacon\Beacon;
+use WP_Rocket\Engine\Admin\Settings\Page;
+use WP_Rocket\Engine\Admin\Settings\Settings;
+use WP_Rocket\Engine\License\API\UserClient;
+use WP_Rocket\Engine\Optimization\DelayJS\Admin\SiteList;
+use WPMedia\PHPUnit\Unit\TestCase;
```
Comment (remyperona) [inline tests/Unit/inc/Engine/Admin/Settings/Page/enableMobileCache.php]:
```suggestion
use WP_Rocket\Tests\Unit\TestCase;
```

File: tests/Unit/inc/Engine/Admin/Settings/Page/enableMobileCache.php [hunk 11]
```diff
@@ -0,0 +1,88 @@
...
+
+/**
+ * @covers \WP_Rocket\Engine\Admin\Settings\Page::enable_mobile_cache
+ */
+class Test_EnableMobileCache extends TestCase {
+    private $page;
+	private $beacon;
+	private $options;
+
+	public function setUp() : void {
```
Comment (remyperona) [inline tests/Unit/inc/Engine/Admin/Settings/Page/enableMobileCache.php]:
```suggestion
	protected function setUp() : void {
```

File: tests/Unit/inc/Engine/Admin/Settings/Page/enableMobileCache.php [hunk 12]
```diff
@@ -0,0 +1,88 @@
...
+			Mockery::mock( SiteList::class ),
+			$template_path,
+			$this->options
+		);
+
+		Functions\when( 'check_ajax_referer' )->justReturn( true );
+	}
+
+	/**
+	 * @dataProvider provideTestData
```
Comment (remyperona) [inline tests/Unit/inc/Engine/Admin/Settings/Page/enableMobileCache.php]:
```suggestion
	 * @dataProvider configTestData
```

File: tests/Unit/inc/Engine/Admin/Settings/Page/enableMobileCache.php [hunk 13]
```diff
@@ -0,0 +1,88 @@
...
+		$this->options->shouldReceive( 'get_options' )
+		              ->once();
+
+		Functions\expect( 'update_option' )->once();
+		Functions\expect( 'wp_send_json_success' )->once();
+	}
+
+	public function provideTestData() {
+		return $this->getTestData( __DIR__, 'enableMobileCache' );
+	}
```
Comment (remyperona) [inline tests/Unit/inc/Engine/Admin/Settings/Page/enableMobileCache.php]:
```suggestion
```

File: tests/Unit/inc/Engine/Admin/Settings/Page/enableSeparateCacheFilesMobile.php [hunk 14]
```diff
@@ -0,0 +1,95 @@
...
+
+use Mockery;
+use Brain\Monkey\Functions;
+use WP_Rocket\Engine\Admin\Database\Optimization;
+use WP_Rocket\Engine\Admin\Beacon\Beacon;
+use WP_Rocket\Engine\Admin\Settings\Page;
+use WP_Rocket\Engine\Admin\Settings\Settings;
+use WP_Rocket\Engine\License\API\UserClient;
+use WP_Rocket\Engine\Optimization\DelayJS\Admin\SiteList;
+use WPMedia\PHPUnit\Unit\TestCase;
```
Comment (remyperona) [inline tests/Unit/inc/Engine/Admin/Settings/Page/enableSeparateCacheFilesMobile.php]:
```suggestion
use WP_Rocket\Tests\Unit\TestCase;
```

File: tests/Unit/inc/Engine/Admin/Settings/Page/enableSeparateCacheFilesMobile.php [hunk 15]
```diff
@@ -0,0 +1,95 @@
...
+
+/**
+ * @covers \WP_Rocket\Engine\Admin\Settings\Page::enable_separate_cache_files_mobile
+ */
+class Test_EnableSeparateCacheFilesMobile extends TestCase {
+    private $page;
+	private $beacon;
+	private $options;
+
+	public function setUp() : void {
```
Comment (remyperona) [inline tests/Unit/inc/Engine/Admin/Settings/Page/enableSeparateCacheFilesMobile.php]:
```suggestion
	protected function setUp() : void {
```

File: tests/Unit/inc/Engine/Admin/Settings/Page/enableSeparateCacheFilesMobile.php [hunk 16]
```diff
@@ -0,0 +1,95 @@
...
+			Mockery::mock( Optimization::class ),
+			Mockery::mock( UserClient::class ),
+			Mockery::mock( SiteList::class ),
+			$template_path,
+			$this->options
+		);
+	}
+
+	/**
+	 * @dataProvider provideTestData
```
Comment (remyperona) [inline tests/Unit/inc/Engine/Admin/Settings/Page/enableSeparateCacheFilesMobile.php]:
```suggestion
	 * @dataProvider configTestData
```

File: tests/Unit/inc/Engine/Admin/Settings/Page/enableSeparateCacheFilesMobile.php [hunk 17]
```diff
@@ -0,0 +1,95 @@
...
+
+		$this->options->shouldReceive( 'get_options' )
+		              ->once();
+
+		Functions\expect( 'update_option' )->once();
+	}
+
+	public function provideTestData() {
+		return $this->getTestData( __DIR__, 'enableSeparateCacheFilesMobile' );
+	}
```
Comment (remyperona) [inline tests/Unit/inc/Engine/Admin/Settings/Page/enableSeparateCacheFilesMobile.php]:
```suggestion
```

**Review by remyperona — CHANGES_REQUESTED** [review]

**Review by remyperona — APPROVED** [review]

**Review by CrochetFeve0251 — APPROVED** [review]

**Comment by Mai-Saad** [comment]
@jeawhanlee Thanks for the PR. 
Exploratory testing notes so far: 

- [ ] We have a problem in the following scenario1: (preload not working when enabling mobile cache in a specific scenario)

1- use 3.13.4
2- enable separate mobile cache
3- disable mobile cache
4- update
5- enable mobile cache => preload didn't clear the table and even visiting the mobile page won't create a mobile cache file
Notes:
- Manual clear and preload cache after 5 works correctly
- if before the update we disabled separate mobile cache too, at step 5 , preload will set table to pending
- after we update, preload won't set to pending till we refresh admin, is this expected or may be it's because I am using the snippet? ===> As per discussion, we need to recheck this scenario with a real update using the internal release



- [x] Another problem with preload here in scenario 2: 
1- 3.13.4 active version
2- enable mobile cache only => preload done
3- update and refresh admin => preload set to pending and once done, no mobile cache files  (even clear and preload will load only desktop)
Notes:
- preload is not working while updating in this case i.e only preload desktop
- if we made a manual visit to mobile before preload is done, mobile files will be added but not for home
clear and preload will load only the desktop in this case
- Fresh install is preloading both mobile and desktop
- update while mobile cache and separate mobile cache both on/off , will preload both mobile and desktop
so the problem is mainly updating while the mobile cache is on and the separate  mobile cache is off

- [ ] Scenario3:
1. install and activate 3.16
2. import settings where the  mobile cache is enabled while the separate mobile cache is off => preload will load only the desktop in this case (Note: the same will occur if imported file has only separate mobile cache on while mobile cache off or both mobile cache and separate mobile cache are off)

@jeawhanlee  can you please check 2, 3 :pray: 

@piotrbak Mobile cache option at tools tab is just before the dynamic lists, is this order or it shall be at the end of page?

**Comment by piotrbak** [comment]
@Mai-Saad That's fine with the order

**Review by remyperona — APPROVED** [review]

**Comment by Mai-Saad** [comment]
@jeawhanlee Thank you for the update scenario 2 is fixed.
1- Scenario 1 mentioned above still not working + **even clear and preload won't preload mobile or reenable preload feature**
2- Regarding import, it's fixed if the imported file has a mobile cache on, separate off, or both off. however if the mobile cache is off and the separate cache is on, then we enable mobile cache at tools, mobile won't be preloaded **even with clear and preload or re-enable preload feature** 

Note: in both 1,2 , if we changed permalinks, mobile version will be preloaded. However, clear and preload cache after that won't preload mobile

@jeawhanlee can you please check. @piotrbak what do you think?

**Review by remyperona — APPROVED** [review]

**Comment by Mai-Saad** [comment]
@jeawhanlee Thanks for the update.
Now Import settings where:
- mobile cache on and separate cache off or both mobile and separate cache are on => preload will occur for both desktop and mobile
- mobile cache off and separate mobile off => When click enable mobile cache, the desktop will be preloaded but not mobile **(clear and preload will preload both desktop and mobile)**
- mobile cache off and separate mobile on =>  When click enable mobile cache, nothing happens in cache ( note: on trunk if we enable/disable mobile cache while separate mobile is on, nothing will be preloaded), however, **if we clear and preload both mobile and desktop will be preloaded**

And update where:
- mobile cache on and separate cache off or both mobile and separate cache are on => **we need to refresh admin so preload works**, it will preload both desktop and mobile
- mobile cache off and separate mobile cache off then click enable mobile cache => cache is preloaded for both mobile and desktop
- mobile cache off but separate on then click enable mobile cache nothing preloaded **till manually clear and preload cache** (after update, need to refresh so that the preload occurs)

@piotrbak Since manual clear and preload is the workaround when no automatic preload happens, do you think further improvement is needed here or shall we proceed with test execution as is for now?

**Comment by Mai-Saad** [comment]
@jeawhanlee @Tabrisrp Thanks for the updates. Please find exploratory test notes below (WIP)

-  Fresh install and activate the PR => only desktop is preloaded while mobile is not preloaded although it should (for retest, we need to retest with WP sitemap, using SEO plugin and custom sitemap)=> **update** after retesting this with zip then with the branch it works
